x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -482,7 +482,13 @@ private void preAnalyzeIndices(
                     indexExpressionToResolve,
                     result.fieldNames,
                     requestFilter,
-                    listener.map(indexResolution -> result.withIndexResolution(indexResolution))
+                    listener.delegateFailure((l, indexResolution) -> {
+                        if (configuration.allowPartialResults() == false && indexResolution.getLocalResolutionFailure() != null) {
+                            l.onFailure(indexResolution.getLocalResolutionFailure().getException());
+                        } else {
+                            l.onResponse(result.withIndexResolution(indexResolution));
+                        }
+                    })
```

Comment:
This change completes the listener early, immediately after index resolution with a failure if there are failures